Improbable/Schema/Improbable.Schema.Bundle/FieldDefinitionExtensions.cs

```diff
@@ -0,0 +1,296 @@
+                FieldType.List => field.ListType.InnerType.ValueTypeSelector == ValueType.Primitive,
+                FieldType.Map => throw new InvalidOperationException(),
+                FieldType.Singular => field.SingularType.Type.ValueTypeSelector == ValueType.Primitive,
+                _ => throw new ArgumentOutOfRangeException()
+            };
+        }
+
+        public static bool IsPrimitive(this FieldDefinition field, PrimitiveType type)
+        {
+            return field.TypeSelector switch
+            {
+                FieldType.Option => field.OptionType.InnerType.ValueTypeSelector == ValueType.Primitive && field.OptionType.InnerType.Primitive == type,
+                FieldType.List => field.ListType.InnerType.ValueTypeSelector == ValueType.Primitive && field.ListType.InnerType.Primitive == type,
+                FieldType.Map => throw new InvalidOperationException(),
+                FieldType.Singular => field.SingularType.Type.ValueTypeSelector == ValueType.Primitive && field.SingularType.Type.Primitive == type,
+                _ => throw new ArgumentOutOfRangeException()
+            };
+        }
+
+        public static bool IsCustomType(this FieldDefinition field)
+        {
+            return field.TypeSelector switch
+            {
+                FieldType.Option => field.OptionType.InnerType.ValueTypeSelector == ValueType.Type,
+                FieldType.List => field.ListType.InnerType.ValueTypeSelector == ValueType.Type,
+                FieldType.Map => throw new InvalidOperationException(),
+                FieldType.Singular => field.SingularType.Type.ValueTypeSelector == ValueType.Type,
+                _ => throw new ArgumentOutOfRangeException()
+            };
+        }
+
+        public static bool IsCustomType(this FieldDefinition field, string qualifiedTypeName)
+        {
+            return field.TypeSelector switch
+            {
+                FieldType.Option => field.OptionType.InnerType.ValueTypeSelector == ValueType.Type && field.OptionType.InnerType.Type == qualifiedTypeName,
+                FieldType.List => field.ListType.InnerType.ValueTypeSelector == ValueType.Type && field.ListType.InnerType.Type == qualifiedTypeName,
+                FieldType.Map => throw new InvalidOperationException(),
+                FieldType.Singular => field.SingularType.Type.ValueTypeSelector == ValueType.Type && field.SingularType.Type.Type == qualifiedTypeName,
+                _ => throw new ArgumentOutOfRangeException()
+            };
+        }
+
+        public static bool IsSingular(this FieldDefinition field)
+        {
+            return field.TypeSelector == FieldType.Singular;
+        }
+
+        public static bool IsSingularPrimitive(this FieldDefinition field)
+        {
+            if (!field.IsSingular())
+            {
+                throw new InvalidOperationException("Field type is not singular");
```

Comment:
These exceptions should try to provide as much detail as possible for the field that caused the errors.

Reply:
These are all going to indicate logic errors in a code generator rather than user input from schema errors, but I'll see what I can do.

Reply:
The extra information helps when debugging potential logic errors though

Reply:
Looking at it more, I don't think there's much more to add other than the type it actually is, which I'll do.
You have the callstack, and you've messed up your type matching and are calling `IsSingular`, etc on a field that's not. It's really all in your code.

Reply:
All info is useful 😄